File: StriverA2Z/JavaScript&TypeScript/src/BinaryTrees/boundaryTraversal.ts
import {arrayToBTLevelOrder, BTNode} from "./BinaryTreeOperations";

function isLeaf(node: BTNode) {
	return node.left === null && node.right === null;
}

function leftBoundary(root: BTNode, result: number[]) {
	let cur = root.left;
	while (cur) {
		if (!isLeaf(cur)) result.push(cur.data);
		if (cur.left) cur = cur.left; else cur = cur.right;
	}
}

function rightBoundary(root: BTNode, result: number[]) {
	let cur = root.right;
	const temp = []
	while (cur) {
		if (!isLeaf(cur)) temp.push(cur.data);
		if (cur.right) cur = cur.right; else cur = cur.left;
	}
	for (let i = temp.length - 1; i >= 0; i--) {
		result.push(temp[i]);
	}
}

function leafBoundary(root: BTNode, result: number[]) {
	if (isLeaf(root)) {
		result.push(root.data);
		return;
	}
	if (root.left) leafBoundary(root.left, result);
	if (root.right) leafBoundary(root.right, result);
}

function boundaryTraversal(root: BTNode) {
	let result = [];
	if (root === null) return result;
	if (!isLeaf(root)) result.push(root.data);
	leftBoundary(root, result);
	leafBoundary(root, result);
	rightBoundary(root, result);
	return result;
}

let root = arrayToBTLevelOrder([1,2,3,4,5,6,7]);
console.log(boundaryTraversal(root))

File: StriverA2Z/Java/BinaryTrees/BoundaryTraversal.java
import java.util.ArrayList;

public class BoundaryTraversal {
    public static void main(String[] args) {
        Node root = BinaryTreeOperations.arrayToBTLevelOrder(new Integer[]{1, 2, 3, 4, 5, 6, 7});
        ArrayList<Integer> result = new ArrayList<>();
        boundaryTraversal(root, result);
        for (Integer i : result) {
            System.out.print(i + " ");
        }
    }

    public static boolean isLeaf(Node root) {
        return root.left == null && root.right == null;
    }

    public static void leftBoundary(Node root, ArrayList<Integer> result) {
        Node cur = root.left;
        while (cur != null) {
            if (!isLeaf(cur)) result.add(cur.data);
            if (cur.left != null) cur = cur.left;
            else cur = cur.right;
        }
    }

    public static void rightBoundary(Node root, ArrayList<Integer> result) {
        Node cur = root.right;
        ArrayList<Integer> temp = new ArrayList<>();
        while (cur != null) {
            if (!isLeaf(cur)) temp.add(cur.data);
            if (cur.right != null) cur = cur.right;
            else cur = cur.left;
        }
        for (int i = temp.size() - 1; i >= 0; i--) {
            result.add(temp.get(i));
        }
    }

    public static void leafNode(Node root, ArrayList<Integer> result) {
        if (isLeaf(root)) {
            result.add(root.data);
            return;
        }
        if (root.left != null) leafNode(root.left, result);
        if (root.right != null) leafNode(root.right, result);
    }

    public static void boundaryTraversal(Node root, ArrayList<Integer> result) {
        if (!isLeaf(root)) result.add(root.data);
        leftBoundary(root, result);
        leafNode(root, result);
        rightBoundary(root, result);
    }
}
